fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt

```diff
@@ -233,6 +233,13 @@ internal object AppStoreReducer {
         )
 
         is AppAction.ShoppingSheetStateUpdated -> state.copy(shoppingSheetExpanded = action.expanded)
+
+        is AppAction.RecordPendingFxSuggestImpression -> state.copy(pendingFxSuggestImpressions = state.pendingFxSuggestImpressions + action.impression)
+        is AppAction.ForgetPendingFxSuggestImpression -> state.copy(pendingFxSuggestImpressions = state.pendingFxSuggestImpressions - action.impression)
+        is AppAction.ForgetAllPendingFxSuggestImpressions -> state.copy(pendingFxSuggestImpressions = emptySet())
+
+        is AppAction.SearchEngagementFinished -> state
+        is AppAction.SearchEngagementAbandoned -> state
```

Comment:
Oh, yes! When I wrote this part, I think I missed that the middleware can run _before_ the state is updated, because it controls when `next()` is called—so when it sees a `SearchEngagementFinished` action, it can capture the pending impressions, and send the Glean ping, before the reducer sees that action and clears them out. Thanks!